### ISSUE TITLE
libykcs11.dylib not working with macOS 10.12.5

### ISSUE THREAD

**Issue by grantho** [Yubico/yubico-piv-tool#118]
Hi, my Yubikey no longer seems to be working with macOS 10.12.5:

When I insert my yubiKey and run "ssh-add -s /usr/local/lib/libykcs11.dylib", I get the following prompts (after entering the correct PIN):

Enter passphrase for PKCS#11: 
Could not add card "/usr/local/lib/libykcs11.dylib": agent refused operation

Any thoughts on how to fix this?

**Comment by a-dma** [comment]
Debug information would be helpful.
Please have a look at #88 for how to enable some debugging and possible things to try.

**Comment by paul-pearce** [comment]
I was able to replicate @grantho's issue under 10.12.5. On my end, it's an issue with `ssh-agent -P, PKCS_WHITELIST`

--

By default the Mac 10.12 `ssh-agent` will only accept libraries located in `/usr/local/lib` or `/usr/lib/`, based on its `-P` invocation. See https://github.com/OpenSC/OpenSC/issues/1008 for more discussion.

If you install `yubico-piv-tool` via `brew install yubico-piv-tool` (which I believe @grantho is, given the path in this issue, and our private conversations) the file located at `/usr/local/lib/libykcs11.dylib` is actually symlink outside of `/usr/local/lib` which violates the `PKCS_WHITELIST`, generating the error above.

This resolves the error.
```
cp /usr/local/lib/libykcs11.dylib /usr/local/lib/libykcs11_NOTALINK.dylib
ssh-add -s /usr/local/lib/libykcs11_NOTALINK.dylib
```

The error is laughably vague. :\

**Comment by paul-pearce** [comment]
I opened an issue with homebrew to see if they can copy the library instead of symlink it (see issue above). I've never interacted with that project. I'm not sure how they work or if such an action is possible. If you all have done work to add the formula, perhaps you can jump in?

**Comment by grantho** [comment]
Confirmed that @paul-pearce 's fix works. Thanks!

**Comment by paul-pearce** [comment]
If folks care about this issue, in the next 2 days I'd head over to the homebrew bug above and bump it. They keep closing it as stale, despite my protesting.

**Comment by micolous** [comment]
I see you've hit Homebrew's perpetual bug bankruptcy problem. :(

Looking at the different options here, as someone else who has tried to get this working for themselves:

1. Changing the `ssh-agent` launchd configuration violates SIP.  We should consider this a non-starter.

2. [Homebrew folks suggest using _their_ OpenSSH](https://github.com/Homebrew/homebrew-core/issues/16385#issuecomment-353788597).

   While I haven't tried this, a cursory glance suggests it _probably wouldn't work_, given [that they don't appear to carry any more patches than Apple does](https://github.com/Homebrew/homebrew-core/blob/527d7873b1da39903413984cace756f8245f54d9/Formula/openssh.rb#L23-L37).

   It would be possible to fix their OpenSSH package to allow Homebrew paths by default, but you still need to get their `ssh-agent` running by default instead of Apple's (which is tedious, but such a replacement is also used by the GnuPG-SSH instructions for YK on OSX).

3. It seems like it's difficult to fix this in the Homebrew side, because it would require special-casing file installation for a single package, and you would lose `brew (un)link` support.

   Unfortunately, this makes the PKCS11 modules included with Homebrew's `yubico-piv-tool` package _completely useless for use with OpenSSH_.  The rest of the package is fine, so it's difficult to argue that they should remove it.

[The Yubico documentation does mention adding some links for `opensc_pkcs11.so`](https://developers.yubico.com/PIV/Guides/SSH_with_PIV_and_PKCS11.html), which I _did_ read, but I _didn't_ do anything about it because there was _already_ a symlink in place.  I didn't click until later that it was proposing to create a _hard link_ instead of a symlink, and certainly [the `-P` flag wasn't mentioned](https://bugs.chromium.org/p/project-zero/issues/detail?id=1009).

I found operating _without_ an agent to work reasonably well, but you do have to type in the PIN every time.  When using an agent, it works once only, and then you need to reload the module (which _also_ requires entering the PIN again).

However, the agent is required for things like `git` over SSH, which is something I do a lot of. :(

#158 has a quite a different problem to this bug, but has a somewhat misleading title.

**Comment by rts-rob** [comment]
From reading this it appears the current status is "Yubikeys don't work with git over SSH on macOS" since `ssh-add` fails no matter what you try?

**Comment by chaimatwork** [comment]
Any updates on this? I have one Yubikey working just fine on Mojave 10.14.6 with OpenSC 0.19.0. Another one on 10.14.? (doesn't say) with the same exact version of OpenSC is failing. Instead of symlink I am using a copy of the file.

Is this a problem with OpenSSH? I cannot SSH anywhere.

**Comment by sammy007** [comment]
These clowns must resolve all issues and make it smooth on OSX, we have paid money for their toys, annoying. No up to date manuals, workarounds, must google random junk article, what a scam.

**Comment by qpernil** [comment]
The current workaround for this is to use ykcs11 directly from ssh with the -I option.

**Comment by qpernil** [comment]
After a little digging I suspect that the fact that while that libykcs11.1.dylib and libykpiv.1.dylib are both under /usr/local/lib the openssl1.1 lib they use (/usr/local/opt/openssl@1.1/lib/libcrypto.1.1.dylib from homebrew) is not, and this might be why it fails the ssh-agent's whitelist check.
(Using /usr/local/lib/libykcs11.dylib should be OK since ssh-agent resolves links before checking)

**Comment by qpernil** [comment]
This seems to work now with Catalina 10.15.4. Not sure if this worked before.

ssh-agent -P "/*" zsh

 (The above starts a new ssh-agent, then a new zsh with the ssh-agent's env variables set up)

ssh-add -s /usr/local/lib/libykcs11.dylib

And we're up. I know this isn't a fix for the original issue, but at least it's a workable solution that is more convenient than using ssh without an agent.

**Comment by qpernil** [comment]
I'm closing this issue now since it's not really an issue with ykcs11.

**Comment by LnxGnome** [comment]
Not sure if this is the OP problem, but ssh-add would give me the _agent refused operation_ error.  It was due to a missing dylib in /usr/local/lib that was not evident when running ssh-add, but it come up when using ssh-keygen -D.
`dependent dylib '@loader_path/libcrypto.1.0.0.dylib' not found for '/usr/local/lib/libykcs11.1.dylib'`
I ended up copying yubico-piv-tool-2.0.0-mac/lib/lib*dylib to /usr/loca/lib and it solved the ssh-add issue for me.
